File: handlers/get_product_inventory.zig
const std = @import("std");
const t = @import("../prelude.zig");

pub const Status = enum { ok, not_found };

pub const Prefetch = struct { product: ?t.ProductRow };

pub const Context = t.HandlerContext(Prefetch, t.Operation.EventType(.get_product_inventory), t.Identity, Status);

pub const route_method = t.http.Method.get;
pub const route_pattern = "/products/:id/inventory";

// [route] .get_product_inventory
// match GET /products/:id/inventory
pub fn route(method: t.http.Method, raw_path: []const u8, body: []const u8) ?t.Message {
    _ = method;
    _ = body;
    const params = t.match_route(raw_path, route_pattern) orelse return null;
    const id = t.stdx.parse_uuid(params.get("id").?) orelse return null;
    return t.Message.init(.get_product_inventory, id, 0, {});
}

// [prefetch] .get_product_inventory
pub fn prefetch(storage: anytype, msg: *const t.Message) ?Prefetch {
    return .{ .product = storage.query(t.ProductRow,
        "SELECT id, name, description, price_cents, inventory, version, active FROM products WHERE id = ?1;",
        .{msg.id}) };
}

// [handle] .get_product_inventory
pub fn handle(ctx: Context, db: anytype) t.HandleResult {
    _ = db;
    const product = ctx.prefetched.product orelse
        return .{ .status = .not_found };
    if (!product.active) return .{ .status = .not_found };
    return .{};
}


// [render] .get_product_inventory
pub fn render(ctx: Context) []const u8 {
    return switch (ctx.status) {
        .not_found => "Product not found",
        .ok => render_ok(ctx),
    };
}

fn render_ok(ctx: Context) []const u8 {
    const product = ctx.prefetched.product.?;
    var pos: usize = 0;
    pos += t.html.raw(ctx.render_buf[pos..], "inventory: ");
    pos += t.html.u32_decimal(ctx.render_buf[pos..], product.inventory);
    return ctx.render_buf[0..pos];
}

File: examples/ecommerce-ts/handlers/get_product_inventory.ts
import type { RouteRequest, RouteResult, PrefetchMessage, PrefetchDb, HandleContext, RenderContext } from "tiger-web";

// [route] .get_product_inventory
// match GET /products/:id/inventory
export function route(req: RouteRequest): RouteResult | null {
  return { operation: "get_product_inventory", id: req.params.id };
}

// [prefetch] .get_product_inventory
export async function prefetch(msg: PrefetchMessage, db: PrefetchDb) {
  const product = await db.query("SELECT id, name, description, price_cents, inventory, version, active FROM products WHERE id = ?1", msg.id);
  return { product };
}

// [handle] .get_product_inventory
export function handle(ctx: HandleContext): string {
  if (!ctx.prefetched.product) return "not_found";
  if (!ctx.prefetched.product.active) return "not_found";
  return "ok";
}

// [render] .get_product_inventory
export function render(ctx: RenderContext): string {
  switch (ctx.status) {
    case "ok": return `<div class="inventory">${ctx.prefetched.product.inventory}</div>`;
    case "not_found": return '<div class="error">Product not found</div>';
  }
  throw new Error("unreachable: " + ctx.status);
}
